daemon/qrexec-client.c

```diff
@@ -749,33 +780,37 @@ int main(int argc, char **argv)
         usage(argv[0]);
     }
 
-    if ((strcmp(domname, "dom0") == 0 || strcmp(domname, "@adminvm") == 0) &&
-            !connect_existing) {
-        fprintf(stderr, "ERROR: when target domain is 'dom0', -c must be specified\n");
-        usage(argv[0]);
-    }
-
     if (strcmp(domname, "dom0") == 0 || strcmp(domname, "@adminvm") == 0) {
+        assert(src_domain_name);
+        if (src_domain_name == NULL)
+            abort();
+        setenv("QREXEC_REMOTE_DOMAIN", src_domain_name, 1);
+        s = connect_unix_socket(src_domain_name);
         if (connect_existing) {
             msg_type = MSG_SERVICE_CONNECT;
             strncpy(svc_params.ident, request_id, sizeof(svc_params.ident) - 1);
             svc_params.ident[sizeof(svc_params.ident) - 1] = '\0';
-        } else if (just_exec)
-            msg_type = MSG_JUST_EXEC;
-        else
-            msg_type = MSG_EXEC_CMDLINE;
-        assert(src_domain_name);
-        setenv("QREXEC_REMOTE_DOMAIN", src_domain_name, 1);
+        } else {
+            fprintf(stderr, "ERROR: when target domain is 'dom0', -c must be specified\n");
+            usage(argv[0]);
```

Comment:
This condition should be checked before connecting anywhere.

Reply:
This is actually the same check that the previous code had.